### DISCUSSION TITLE
VPN issue

### DISCUSSION THREAD

**Discussion by 26zl** [utmapp/UTM#6792] [UTM Help (macOS)]
Hi!

I’m using a Mac Mini M2 (latest os) with Mullvad VPN (latest), and I’m running a Kali Linux VM through UTM (latest version). The objective is to route all traffic from the VM through the host’s VPN (Mullvad), ensuring that my real IP do not get exposed. However, I’ve encountered issues with this setup:

Only “Bridged” mode provides internet access, but it bypasses the VPN, showing my real IP in the VM.

I’ve tried all available network modes and configurations in UTM with no success in routing the VM traffic through Mullvad on the host i do not get any internet at all with these modes enabled with Mullvad on.

I’ve tested multiple Linux distributions, and the issue persists across them all, so not only kali.

Network Modes

Shared Network (NAT):
No connection at all

Bridged Mode:
This mode allowed the VM to access the internet, but it bypassed the VPN entirely, showing my real IP in the VM while the Mac itself was connected through Mullvad VPN.

Host Only:
Blocks internet access for the VM, so it wasn’t suitable for our purposes.

Emulated VLAN:
I've seen this being the problem solver for most people with the same problem, but it did not work for me unfortunately.

Other settings:

Virtio Network Card:
I tested both virtio-net-pci and virtio-net-device as network card options. Both configurations yielded the same result, with no difference in routing behavior.

Mullvad Settings:
Enabled in Mullvad’s settings. This option should, in theory, allow traffic from local network sources (such as the VM) to route through the VPN. It’s worth mentioning that on Windows, this setting must be enabled to allow similar setups, so it was enabled here on macOS as well to ensure compatibility with what I’m trying to achieve.

Firewall: MacOS firewall is enabled, but UTM and Mullvad VPN are allowed through, so they should not be blocked.

Isolate Guest from Host:
I enabled and disabled Isolate Guest from Host to test if it would enforce routing through the VPN by preventing direct communication between the VM and the host. However, this setting made no difference in the routing behavior for the VM.

Manual Static Routes:
I attempted to add a route within the VM to direct traffic through the host’s VPN interface (utun0). This resulted in the error “no such device”, since utun0 exists only on the host and is not visible within the VM.

I love UTM and genuinely appreciate the software you've built. I’ve had nothing but positive experiences with the program otherwise, and I’m hopeful someone might be able to help me resolve this specific issue with VM routing on macOS. Any advice or guidance would be greatly appreciated!

**Comment by bendvc** [comment]
I am having the exact same issue. @26zl have you made any progress on this or have gotten any work arounds going? My corporate network profiles aren't allowing me to create any type of host proxies to route traffic through, but maybe that would work for you.

**Reply by 26zl** [reply]
Unfortunately, no. For now, I’m using VMware Fusion 13 Pro (which is now free) for Kali and other ARM-based systems. It works very well, and I haven’t encountered any issues. VPN/network problems are also resolved when using VMware.

**Comment by frank47111** [comment]
Same issue here.
We need the guest OS to share network with the host in MacOS. However, the default in shared network is that the guest receives a DHCP IP address and is using this IP to access the network and only shares the NIC of the host. This traffic is (in my case) correctly sent into my VPN, however since the IP is a different one, return traffic does not come back.
What we need is the guest to share the IP and MAC address of the host (use NAT)

I other virtualization systems (like Vmware Fusion) this is possible (but they cannot emulate another architecture). However, I did not find this in UTM or Qemu documentation.

How is this possible with UTM?
